lib/vuln.js

```diff
@@ -94,10 +94,15 @@ class Vuln {
   }
 
   toJSON () {
-    // sort so that they're always in a consistent order
+    const edgesIn = [...this.nodes.values()].flatMap(n => [...n.edgesIn])
+    const isDirect = edgesIn.some(e => e.from.isProjectRoot)
+    const types = [...new Set(edgesIn.map(a => a.type))].sort()
```

Comment:
A node might be "optional", but not have any optional edgesIn.  Or it might have optional edgesIn, but not actually be optional.

For example:

```
root -> (a, b) OPTIONAL(c)
a -> OPTIONAL(d)
b -> (d)
c -> (e)
```

In this dep graph,

- `a` is not an optional dep, but has an optional dep on `d`
- but `b` has a production dep on `d`, so `d` is _not_ optional
- `c` is an optional dep of the root project, and nothing else, so it's optional.
- `e` is a production dep, but a production dep _of_ an optional dep, so it is optional.

Is the intention to show all flags on _any_ nodes in the set, or _all_ nodes in the set?  Eg, if there's one node in the tree that is optional, and another that isn't, should `types` include `'optional'`?

```suggestion
    // this will add the flag to `types` if _any_ dep has the flag set
    const allTypes = ['dev', 'devOptional', 'optional', 'peer']
    const types = new Set()
    // if the semantic should be _all_ deps must have it, do this instead:
    // const types = new Set(allTypes)
    let isDirect = false
    for (const node of this.nodes.values()) {
      for (const t of allTypes) {
        if (node[t])
          types.add(t)
        // if the semantic should be _all_ deps must have the flag set, do this instead:
        // if (!node[t])
        //   types.delete(t)
      }
      // it's a direct dep if the root project or any workspace depends directly on it
      // since workspace projects are "ours" in the same sense as the root project
      // only need to check if we haven't already found one.
      if (!isDirect) {
        for (const { from } of node.edgesIn) {
          if (from.isProjectRoot || from.isWorkspace) {
            isDirect = true
            break
          }
        }
      }
    }
```

Reply:
I actually only needed dev, so I'm reluctant to include this much complexity here for figuring out types.
I'm going to split this PR so I can finish isDirect and get it merged and use it, then work on types.

Reply:
Extracted here: https://github.com/npm/arborist/pull/323